sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,681 @@
+
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(float start, float *ptr) {
+  // Get the ID of this kernel iteration.
+  size_t id = syclex::this_kernel::get_id();
+
+  ptr[id] = start + static_cast<float>(id);
+}
+
+void main() {
+  sycl::queue q;
+  sycl::context ctxt = q.get_context();
+
+  // Get a kernel bundle that contains the free function kernel "iota".
+  auto exe_bndl =
+    syclex::get_kernel_bundle<iota, sycl::bundle_state::executable>(ctxt);
+
+  // Get a kernel object for the "iota" function from that bundle.
+  sycl::kernel k_iota = exe_bndl.ext_oneapi_get_kernel<iota>();
+
+  float *ptr = sycl::malloc_shared<float>(NUM, q);
+  q.submit([&](sycl::handler &cgh) {
+    // Set the values of the kernel arguments.
+    cgh.set_args(3.14f, ptr);
+
+    cgh.parallel_for({NUM}, k_iota);
+  }).wait();
+}
+```
+
+=== Free function kernels which are templates or overloaded
+
+A free function kernel may be defined as a function template.
+It is also legal to define several overloads for a free function kernel.
+The following example demonstrates how to get a kernel identifier in such
+cases.
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+template<typename T>
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(T start, T *ptr) {
+  // ...
+}
+
+SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL
+void ping(float *x) {
+  // ...
+}
+
+SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL
+void ping(int *x) {
+  // ...
+}
+
+int main() {
+  // When the free function kernel is templated, pass the address of a
+  // specific instantiation.
+  sycl::kernel_id iota_float = syclex::get_kernel_id<iota<float>>();
+  sycl::kernel_id iota_int = syclex::get_kernel_id<iota<int>>();
+
+  // When there are multiple overloads of a free function kernel, use a cast
+  // to disambiguate.
+  sycl::kernel_id ping_float = syclex::get_kernel_id<(void(*)(float))ping>();
+  sycl::kernel_id ping_int = syclex::get_kernel_id<(void(*)(int))ping>();
+}
+```
+
+
+== Issues
+
+* We're pretty sure that we want to define some syntax that allows a free
+  function kernel to be enqueued using the APIs defined in
+  link:../proposed/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions], but we haven't settled on the exact API
+  yet.
+  One option is like this:
++
+```
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(float start, float *ptr) { /*...*/  }
+
+int main() {
+  sycl::queue q;
+  float *ptr = sycl::malloc_shared<float>(N, q);
+  sycl::parallel_for<iota>(q, {N}, 1.f, ptr);
+}
+```
++
+Another option is like this:
++
+```
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(float start, float *ptr) { /*...*/  }
+
+int main() {
+  sycl::queue q;
+  float *ptr = sycl::malloc_shared<float>(N, q);
+  sycl::parallel_for(q, {N}, kfp<iota>, 1.f, ptr);
+}
+```
++
+Where `kfp` would have some nicer name.
+
+* Can the front-end define a trait like this, which returns true only if the
+  address is a free function kernel:
++
+```
+template <auto *Func>
+bool is_kernel_function_v;
+```
++
```

Comment:
I would expect the trait to have to be backed by some kind of compiler magic since, as far as I know, there is no way in standard C++ to query whether a function was declared with a particular attribute. If the trait is used in host code, that would exclude support for SYCL-unaware host compilers. Presumably, the macros would expand to nothing for such host compilers anyway. Providing an implementation for device-side compilation should not pose any difficulty though.

Reply:
The main use case I have in mind requires this trait to be available in the host compilation pass, so it is only useful if we can make this work.

I think it does require compiler magic, so it probably cannot be done with a third-party host compiler.  The trait would still be useful if it only works when our clang is used for the host compiler pass, though.  This just means we will get better error messages when people use our clang as the host compiler.

It's *almost* possible to implement this trait in the integration header, but I can't think how to make it work in all cases.  For the cases where the kernel function is defined at namespace scope, the integration header can look like this:

```
#include <type_traits>

void foo();  // Compiler generated forward declaration of application's kernel function

// Base trait definition
template<auto *Func> struct is_kernel_function : std::false_type {};

// Specializations for each of the application's kernel functions
template<> struct is_kernel_function<foo> : std::true_type {};
```

However, I can't think to to make this work for the case when the kernel is defined as a static member function of a class.  I also can't think how to make it work if the kernel takes a parameter of some user-defined type that we can't forward declare in the integration header.  Therefore, I think we would need compiler magic to solve this.

Reply:
I added the traits to the spec in ae6f7ede568821463bed0744594db5a442289f1d.  I also defined special restrictions for implementations that use an integration header in 0e1be43b440e19378bddf6c9c09cfc5c1160440f.